### PR TITLE
fix for extract zip from gcs

### LINKED ISSUE #3282: Extract zip/archive files in GCS download
/kind bug

**What steps did you take and what happened:**

Created InferenceService with `storageUri` as a zip archive, e.g. `gs://bucket/folder/archive.zip` - no model files found.

Created InferenceService with `storageUri` as a folder containing a single archive, e.g. `gs://bucket/folder` - archive found and downloaded, but not extracted.

Used `kserve.storage.storage.Storage._download_gcs` directly, with either zip archive or gcs folder containing only a single archive - either downloads nothing, or downloads the zip but doesn't extract it.

**What did you expect to happen:**

I expected the zip to be downloaded and extracted.

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]

**Anything else you would like to add:**

https://github.com/kserve/kserve/pull/1836/files added code which should do this, but there are a couple of problems:

1. The uri is always used as a prefix, with a [trailing slash added](https://github.com/kserve/kserve/blob/cac5fb1407391eedee284009d31c5a45ef127b23/python/kserve/kserve/storage/storage.py#L250) - so specifying the uri as `gs://bucket/folder/archive.zip` results in any blobs with the prefix `folder/archive.zip/` being downloaded - there are none.
2. The folder itself is counted as a blob (I'm not sure if this is specific to the google cloud storage version I'm using - I created a fresh poetry project and just installed kserve (^0.11.2), and it chose google-cloud-storage 1.44.0), so the `file_count` is incremented for the folder as well which results in `file_count` of 2, so it [doesn't extract the archive](https://github.com/kserve/kserve/blob/cac5fb1407391eedee284009d31c5a45ef127b23/python/kserve/kserve/storage/storage.py#L272). I'm not sure including the folder in the count is the desired behaviour here - possibly this is an indentation issue?

**Environment:**

- Istio Version: 
- Knative Version: 
- KServe Version: 0.11.2
- Kubeflow version:
- Cloud Environment: 
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`): Debian GNU/Linux 10 (buster)

### PR DESCRIPTION
**Pull request by andyi2it**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3282 

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?

**Release note**:

```release-note
NONE

### CODE REVIEW

File: python/kserve/kserve/storage/test/test_storage.py [hunk 1]
```diff
@@ -144,8 +144,10 @@ def test(_):
 def test_mock_gcs(mock_storage):
     gcs_path = 'gs://foo/bar'
     mock_obj = mock.MagicMock()
-    mock_obj.name = 'mock.object'
-    mock_storage.Client().bucket().list_blobs().__iter__.return_value = [mock_obj]
+    mock_obj.name = 'bar/'
+    mock_obj1 = mock.MagicMock()
+    mock_obj1.name = 'bar/mock.object'
+    mock_storage.Client().bucket().list_blobs().__iter__.return_value = [mock_obj, mock_obj1]
```
Comment (yuzisun) [inline python/kserve/kserve/storage/test/test_storage.py]: Can we add more tests for GCS like we have for s3/azure ? not sure what we are asserting here

Comment (spolti) [reply python/kserve/kserve/storage/test/test_storage.py]: IIRC, I saw something related to multimodel, a feature request I guess, wouldn't be better to align with it first?

**Review by spolti — COMMENTED** [review]
the nitpick was a indentation issue :)

**Review by spolti — APPROVED** [review]
/lgtm

**Review by terrytangyuan — APPROVED** [review]
/lgtm

**Pull request merged**